### PR TITLE
Refactor IPC API implementation

### PR DESCRIPTION
**Pull request by Dampfwalze**
This is my go on the IPC API.

The driver events are now expressed by a [`Stream`](https://docs.rs/futures/latest/futures/stream/trait.Stream.html). This also allowes for multiple listeners.

Using the sync API, one can now add many receiver callbacks. The termination is now expressed using a boolean return value. 

```rust
driver.add_event_receiver(|cmd| {
  ...
  if finished {
    false
  } else { 
    true 
  };
});
```

Maybe a subscription object is better?

```rust
let subscription = driver.add_event_receiver(|cmd| { ... });
...
subscription.cancel();
```

(Changes are not fully tested)

### CODE REVIEW

File: rust/driver-ipc/src/client.rs
Comment (MolotovCherry) [inline rust/driver-ipc/src/client.rs:159]: Can we remove this println?

File: rust/driver-ipc/src/client.rs
Comment (MolotovCherry) [inline rust/driver-ipc/src/client.rs:57]: I know I didn't have this in the original code, but I think we should handle the case where this errors out. Basically, what I'm envisioning here is that read errors out, but it's then confusing to the library user on why they aren't receiving any messages (e.g. on their event stream).

The senders should be fine since they'll keep receiving errors if they try to send, but since we still hold onto a broadcast subscriber, the subscribers won't receive a `None`.

I'm thinking that  we should completely drop the broadcast sender on error to solve this.

File: rust/driver-ipc/src/sync/client.rs
Comment (MolotovCherry) [inline rust/driver-ipc/src/sync/client.rs:55]: The changes I mentioned above about doing Result for lagged, should fix this from prematurely terminating because of a lag. (As well as dropping the sender, so it doesn't continue waiting if read prematurely exited)

**Review by MolotovCherry — CHANGES_REQUESTED** [review]
I love where this concept is going. This implementation looks good. Thanks for the contribution!

File: rust/driver-ipc/src/sync/client.rs
Comment (MolotovCherry) [inline rust/driver-ipc/src/sync/client.rs:56]: We also don't know how long `cb` will be executed for, so we should execute this on a blocking thread (maybe a pool?)

File: rust/driver-ipc/src/client.rs
Comment (Dampfwalze) [reply rust/driver-ipc/src/client.rs:121]: This `None` just means to skip this value. This is how `filter_map` is defined.

Comment (MolotovCherry) [reply rust/driver-ipc/src/client.rs:121]: Yeah, this one's fine. Unless we have some reason we want to differentiate which errors are coming through (I can't think of any reason at the moment)

File: rust/driver-ipc/src/driver_client.rs
Comment (Dampfwalze) [reply rust/driver-ipc/src/driver_client.rs:9]: I was thinking, it could be nice to be able to clone the client and work on different states simultaniously, but it might be missleading to use `clone` for this purpose? If we change this, there will be no possibility to do so. 

A usecase might be to "backup" the current state, make changes to the driver and afterwarts revert to the backup.

**Comment by Dampfwalze** [comment]
I was thinking about the possibility to easily revert the drivers state to the persistant state. There could be a method `Client::revert()` or `Client::revert_to_persistant()` that will load back the saved state from the registry.

File: rust/driver-ipc/src/sync/client.rs
Comment (Dampfwalze) [reply rust/driver-ipc/src/sync/client.rs:55]: As mentioned, the stream does not close because of a `Lagged` error, so this should be no problem.

**Comment by MolotovCherry** [comment]
> I was thinking about the possibility to easily revert the drivers state to the persistant state. There could be a method `Client::revert()` or `Client::revert_to_persistant()` that will load back the saved state from the registry.

That's a good suggestion.

What do you think about having the method fetch and convert the persistent state to `Vec<Monitor>` and let the caller do what they want with it? We could keep the one that also reverts the driver in one call as well if it's still desirable

File: rust/driver-ipc/src/client.rs
Comment (MolotovCherry) [reply rust/driver-ipc/src/client.rs]: Whether the callbacks hung or not would be invisible to the user anyways. Still I'm wondering if maybe we should add a flag, callback, or something somewhere to indicate the connection status of the client.

File: rust/driver-ipc/src/driver_client.rs
Comment (MolotovCherry) [reply rust/driver-ipc/src/driver_client.rs:9]: Yeah. I'd rather support only one case (independent states or synced states), since two would just add to confusion. Your use-case is better in the end as well, so my vote would be to not support synced states 

It's still not too clear whether clone would sync states across clients or keep them separate however, so maybe this would be better as a separate method with doc comment to make it clear

**Comment by Dampfwalze** [comment]
> What do you think about having the method fetch and convert the persistent state to `Vec<Monitor>` and let the caller do what they want with it? We could keep the one that also reverts the driver in one call as well if it's still desirable

In any case, there could be a method on the `Client` that deserializes the persistend state: `Client::read_persistent() -> Vec<Monitor>` (also forwarded in `DriverClient`?). 

Then, there can be a method on the `DriverClient` to read this state and write it in its own state: `DriverClient::load_persiststent()`.

But `Client::revert_to_persistant()` is still good to have (also forwarded in `DriverClient`).

**Comment by MolotovCherry** [comment]
> In any case, there could be a method on the `Client` that deserializes the persistend state: `Client::read_persistent() -> Vec<Monitor>` (also forwarded in `DriverClient`?). 
> 
> Then, there can be a method on the `DriverClient` to read this state and write it in its own state: `DriverClient::load_persiststent()`.
> 
> But `Client::revert_to_persistant()` is still good to have (also forwarded in `DriverClient`).

That sounds perfect. I'll add that after this pr is merged

File: rust/driver-ipc/src/sync/client.rs
Comment (Dampfwalze) [reply rust/driver-ipc/src/sync/client.rs:56]: Hmm, good point. Currently the runtime is configured with only one worker thread. It simply does not need more than this. I am not comfortable with allocating more than that, just because it might be that the user blocks the runtime for too long. I much rather prefer to tell the user to not do that. At least for the Dart bindings this does not matter, because it will be dispached asynchronously anyways (aslo, it uses the async api anyways). In case of Python, can we dispatch it asynchronously too? The GIL blocks it from running concurrently anyways.

But it should be just fine, if the runtime is blocked for a few milliseconds (maybe to update some UI). Events won't be dropped, just delayed. And depending on the users use case, it might also be fine, if it is blocked for a few seconds (But the User should at least be aware of that).

Comment (MolotovCherry) [reply rust/driver-ipc/src/sync/client.rs:56]: A comment mentioning it is fine. I think long blocks should be avoided at all costs, but in principle it's not a problem if the user is aware of that. Pretty much any language has the ability to handle this properly as long as the user knows.

Let's add this warning to the doc comment then.

**Comment by Dampfwalze** [comment]
> I would prefer a subscription object instead

I have implemented this now.

One thing to note, there is no good way to cancel the subscription from the callback itself now. Doing this will involve creating some shared state where the subscription is moved into after subscribing, so something like `Arc<Mutex<Option<EventsSubscription>>>`.

Example:

```rust
let shared_sub = Arc::new(Mutex::new(None::<EventsSubscription>));

let sub = client.add_event_receiver({
    let shared_sub = shared_sub.clone();
    move |event| {
        println!("{:?}", event);
        // ...
        shared_sub.lock().unwrap().as_mut().unwrap().cancel();
    }
});

*shared_sub.lock().unwrap() = Some(sub);
```

**Comment by MolotovCherry** [comment]
> > I would prefer a subscription object instead
> 
> I have implemented this now.
> 
> One thing to note, there is no good way to cancel the subscription from the callback itself now. Doing this will involve creating some shared state where the subscription is moved into after subscribing, so something like `Arc<Mutex<Option<EventsSubscription>>>`.
> 
> Example:
> 
> ```rust
> let shared_sub = Arc::new(Mutex::new(None::<EventsSubscription>));
> 
> let sub = client.add_event_receiver({
>     let shared_sub = shared_sub.clone();
>     move |event| {
>         println!("{:?}", event);
>         // ...
>         shared_sub.lock().unwrap().as_mut().unwrap().cancel();
>     }
> });
> 
> *shared_sub.lock().unwrap() = Some(sub);
> ```
> 
> Also: What should happen if the callback panics? Currently the task would just stop without notice. Maybe `EventSubscription::cancel()` could return a `Result`? What would the Error type be then?

I don't feel it matters much here if the callback panics. It does print a message, and the task exiting on panic is expected (similar to what would happen if a thread panicked). Though a note could be added if there's anything we want the user to be clear about. Do you feel it would be useful to catch unwind, mark the panic, resume unwind, and notify on cancel?

I just did some testing and found that the `cb` isn't getting executed at all.

**Comment by Dampfwalze** [comment]
> I just did some testing and found that the `cb` isn't getting executed. The streams are receiving as expected, it seems to stop working at `RUNTIME.spawn` in `EventsSubscription`. A `block_on` does execute it all the way (just a test to see if that was the issue), but `spawn` seems to not

This is wired, because all unit tests pass and they do check if the callback is run or not. I need a bit more context to make sense of the situation here. Maybe the runtime was blocked somehow in your tests? Did you call the API from within the runtime? `RUNTIME.spawn()` does not immediately start execution, it just sends it of to be scheduled for execution.

**Comment by MolotovCherry** [comment]
> This is wired, because all unit tests pass and they do check if the callback is run or not. I need a bit more context to make sense of the situation here. Maybe the runtime was blocked somehow in your tests? Did you call the API from within the runtime? `RUNTIME.spawn()` does not immediately start execution, it just sends it of to be scheduled for execution.
> 
> You can try `tokio::task::yield_now().await` after calling `add_event_receiver()` to give execution back to tokio.

So far from what I've been able to gather, the test `event_receiver_cancel_from_cb()` still passes after a panic is introduced into the callback. The first one however seems to properly fail when doing the same thing, so it applies only to this test. Hmm
```rs
      let sub = client.add_event_receiver({
            let shared_sub = shared_sub.clone();
            move |event| {
                panic!();
            }
        });
```

**Comment by Dampfwalze** [comment]
> So far from what I've been able to gather, the test `event_receiver_cancel_from_cb()` still passes after a panic is introduced into the callback. The first one (`event_receiver()`) properly fails when doing the same thing, so it applies only to this test. Hmm

Yes, this is correct. This makes sense, because when the callback panics, the receiver used to cancel the callback is dropped and closed. If you now call `cancel()`, it returns `false`, which is actually the expected behaviour, because the callback was already "canceled".

**Comment by MolotovCherry** [comment]
Give me a moment to track this original problem down. I'll reply back when I have more context.

File: rust/driver-ipc/src/driver_client.rs
Comment (Dampfwalze) [reply rust/driver-ipc/src/driver_client.rs]: Resolved in 1985a4094028d01c054fc9fe7725c4e939cacdc3

**Comment by MolotovCherry** [comment]
If you'd like to test what I was talking about earlier, here's how:

In the python bindings dir, `lib.rs`, change `receive()` to the following. Compile the program, change the extension to `.pyd`
```rust
    fn receive(&mut self, callback: Option<PyObject>) {
        if let Some(callback) = callback {
            self.client.add_event_receiver(move |cmd| {
                println!("{cmd:?}");
                let EventCommand::Changed(data) = cmd else {
                    unreachable!()
                };

                Python::with_gil(|py| {
                    let state = state_to_pylist(py, &data);
                    let Ok(state) = state else {
                        return;
                    };

                    if let Err(e) = callback.call1(py, (state,)) {
                        e.print(py);
                    }
                });
            });
        }
    }
```
Then run this python:
```python
from vdd import *
c = DriverClient()

def recv(d):
    print(d)

c.receive(recv)
```

File: rust/driver-ipc/src/driver_client.rs
Comment (MolotovCherry) [reply rust/driver-ipc/src/driver_client.rs]: > Resolved in [1985a40](https://github.com/MolotovCherry/virtual-display-rs/commit/1985a4094028d01c054fc9fe7725c4e939cacdc3)

Lgtm!

**Comment by Dampfwalze** [comment]
> Edit 2: Okay I think we got it now. Not exactly surprising (obviously, give the behavior of receivers), but we need to add a comment explaining that the returned value needs to be saved and _not dropped_ until done, otherwise the callback will stop firing. Not binding the value to something will also cause a drop before end of scope. Though obvious, it's an easy gotcha people may accidentally hit.

This is not the behavior I expected nor wanted. I tracked that issue down and fixed it. If the subscription gets dropped, the callback will still run. This behavior is in line with the behavior of `thread::spawn` or `task::spawn`. When you drop the `JoinHandle`, the thread/task will be detached.

**Comment by MolotovCherry** [comment]
The last remaining unresolved comment (below) is already taken care of, isn't it?
> Basically, what I'm envisioning here is that read errors out, but it's then confusing to the library user on why they aren't receiving any messages (e.g. on their event stream).

As far as this, I can take care of handling the error and a callback notification system of sorts / letting user know they got disconnectd.

> The senders should be fine since the caller will receive errors if they try to send, but since we still hold onto a broadcast sender, the subscribers won't receive a None when the stream should be closed.

Iirc, this is already fixed after you changed the `Client` to hold a receiver instead.

---
So I think the previous one can be marked as resolved.

File: rust/driver-ipc/src/client.rs
Comment (Dampfwalze) [reply rust/driver-ipc/src/client.rs]: These are error types I would propose.

```rust
pub mod error {
    use thiserror::Error;
    use super::*;

    /// Error returned from [Client::connect] and [Client::connect_to].
    #[derive(Debug, Error)]
    pub enum ConnectionError {
        #[error("Failed to open pipe: {0}")]
        Failed(io::Error),
    }

    /// Error returned from [send_command]
    #[derive(Debug, Error)]
    pub(super) enum SendCommandError {
        #[error("Failed to encode message: {0}")]
        Encode(serde_json::Error),
        #[error("Failed to send message: {0}")]
        PipeBroken(io::Error),
    }

    /// Error returned from [Client::notify], [Client::remove] and
    /// [Client::remove_all].
    #[derive(Debug, Error)]
    pub enum SendError {
        #[error("Failed to send message: {0}")]
        PipeBroken(io::Error),
    }

    /// Error returned from [Client::request_state].
    #[derive(Debug, Error)]
    pub enum RequestError {
        #[error("Failed to send message (pipe broken): {0}")]
        Send(io::Error),
        #[error("Failed to receive message (pipe broken): {0}")]
        Receive(io::Error),
        #[error("Did not get a response in time ({0:?})")]
        Timeout(Duration),
    }

    // TODO: Client::receive_events()

    /// Error returned from [Client::persist].
    #[derive(Debug, Error)]
    pub enum PersistError {
        #[error("Failed to open registry key: {0}")]
        Open(io::Error),
        #[error("Failed to set registry value: {0}")]
        Set(io::Error),
        #[error("Failed to serialize monitors: {0}")]
        Serialize(serde_json::Error),
    }
}
```

Comment (Dampfwalze) [reply rust/driver-ipc/src/client.rs]: One problem: We cannot easily distribute errors from `send_command()`, since `io::Error` is not copy.

File: rust/driver-ipc/src/client.rs
Comment (MolotovCherry) [inline rust/driver-ipc/src/client.rs:44]: The two client methods are a little misleading, cause even though you can create a named pipe using sync code, it actually requires a reactor to be running. In pure sync code it would result in
`there is no reactor running, must be called from the context of a Tokio 1.x runtime`

We should have both of these be `async` to reduce the possibility anyone would run into the error (though it might be a slightly remote chance since if they're using this they're likely already running tokio in the first place, but regardless)

File: rust/driver-ipc/src/client.rs
Comment (MolotovCherry) [reply rust/driver-ipc/src/client.rs]: > One option is to wrap the event that is send to receivers into a Result. So command_tx would have type Sender<Result<ClientCommand, SomeErrorType>>, receive_events() would return impl Stream<Item = Result<EventCommand, SomeErrorType>>.

I agree with this. It would allow us to send an error to all receivers once the pipe breaks, and we could easily resume `receive_command` again if the client asks to reconnect. The implementation should be real easy to do, and it's clear what has happened since the error was propagated to clients.

> One problem: We cannot easily distribute errors from send_command(), since io::Error is not Clone.

Looks like we can work around this by wrapping it in a `Rc` or `Arc` at least

> But on that note, we should maybe rethink the error system to return specialized errors instead of the general IpcError, since it does not convey what can actually happen. For example, request_state() will never return QueryNotFound error, but the user must still provide a valid path for it.

I also think we should separate these into more specific errors according to the error conditions that can happen (much less superfluous when handling error types for the user).

**Comment by MolotovCherry** [comment]
Lgtm! If you feel it's good to go, I'm fine merging it right now.

**Review by MolotovCherry — APPROVED** [review]

**Comment by MolotovCherry** [comment]
Merged now. Thank you for the time you put into this contribution! <3

**Pull request merged**
